Toolkit/ArcGISToolkit/AR/ArcGISARView.swift

```diff
@@ -0,0 +1,682 @@
+
+        frameCountTimer?.invalidate()
+    }
+    
+    // MARK: Private
+    
+    /// Operations that happen after device tracking has started
+    fileprivate func finalizeStart() {
+        //
+        // hide the camera image if necessary
+        arSCNView.isHidden = !renderVideoFeed
+        
+        //
+        // run the ARSession
+        arSCNView.session.run(arConfiguration, options:.resetTracking)
+        didStartOrFailWithError(nil)
+    }
+
+    /// Adds subView to superView with appropriate constraints
+    ///
+    /// - Parameter subview: the subView to add
+    fileprivate func addSubviewWithConstraints(_ subview: UIView) {
+        // add subview to view and setup constraints
+        addSubview(subview)
+        subview.translatesAutoresizingMaskIntoConstraints = false
+        NSLayoutConstraint.activate([
+            subview.leadingAnchor.constraint(equalTo: self.leadingAnchor),
+            subview.trailingAnchor.constraint(equalTo: self.trailingAnchor),
+            subview.topAnchor.constraint(equalTo: self.topAnchor),
+            subview.bottomAnchor.constraint(equalTo: self.bottomAnchor)
+            ])
+    }
+    
+    /// Start the locationManager with undetermined access
+    fileprivate func startWithAccessNotDetermined() {
+        if (Bundle.main.object(forInfoDictionaryKey: "NSLocationWhenInUseUsageDescription") != nil) {
+            locationManager.requestWhenInUseAuthorization()
+        }
+        if (Bundle.main.object(forInfoDictionaryKey: "NSLocationAlwaysUsageDescription") != nil) {
+            locationManager.requestAlwaysAuthorization()
+        }
+        else{
+            didStartOrFailWithError(ArcGISARView.missingPListKeyError())
+        }
+    }
+    
+    /// Start updating the location and heading via the locationManager
+    fileprivate func startUpdatingLocationAndHeading() {
+        locationManager.startUpdatingLocation()
+        if CLLocationManager.headingAvailable() {
+            locationManager.startUpdatingHeading()
+        }
+    }
+    
+    /// Stop updating location and heading
+    fileprivate func stopUpdatingLocationAndHeading() {
+        locationManager.stopUpdatingLocation()
+        if CLLocationManager.headingAvailable() {
+            locationManager.stopUpdatingHeading()
+        }
+    }
+
+    /// Start the locationManager with denied access
+    fileprivate func startWithAccessDenied() {
+        didStartOrFailWithError(ArcGISARView.accessDeniedError())
+    }
+    
+    /// Start the locationManager with authorized access
+    fileprivate func startWithAccessAuthorized() {
+        startUpdatingLocationAndHeading()
+    }
+    
+    /// Potential notification to the user of an error starting device tracking
+    ///
+    /// - Parameter error: error that ocurred when starting tracking
+    fileprivate func didStartOrFailWithError(_ error: Error?) {
+        if !notifiedStartOrFailure, let error = error {
+            // TODO: present error to user...
```

Comment:
The current thought is that all of the location and heading stuff will be removed from here and replaced with an `AGSLocationDataSource`.  That will simplify a lot of the code and it has it's own mechanism for notifying clients of errors.  If it's OK with you, I'd like to leave this as is (with maybe a note that it will be changing).